adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -103,29 +103,45 @@ public final void start(Future<Void> startFuture) throws Exception {
         } else {
             eventSenderSupplier = (tenant, resultHandler) -> getHonoClient().getOrCreateEventSender(tenant, resultHandler);
             telemetrySenderSupplier = (tenant, resultHandler) -> getHonoClient().getOrCreateTelemetrySender(tenant, resultHandler);
+
+            Future<Void> preStartupFuture = Future.future();
+            preStartup(preStartupFuture);
+            Future<Void> httpServerStartupFuture = Future.future();
+            preStartupFuture.compose(ar -> {
+                bindHttpServer(createRouter(), httpServerStartupFuture);
+                connectToHono(null);
+            }, httpServerStartupFuture);
             
-            Future<Void> beforeHttpStartFuture = Future.future();
-            beforeHttpServerStart(beforeHttpStartFuture);
-            beforeHttpStartFuture.setHandler(ar -> {
-               if (ar.succeeded()) {
-                   bindHttpServer(createRouter(), startFuture);
-                   connectToHono(null);
-               } else {
-                   startFuture.fail(ar.cause());
-               }
+            httpServerStartupFuture.setHandler(ar -> {
```

Comment:
You should connect all of these by means of `Future.compose()`. There are two versions of the `Future.compose()` method, one of them can be used for *middle* steps, the other one for the *last* step.